skbio/stats/distance/_base.py

```diff
@@ -369,6 +369,9 @@ def filter(self, ids, strict=True):
         MissingIDError
             If an ID in `ids` is not in the object's list of IDs.
         """
+        if tuple(self._ids) == tuple(ids):
```

Comment:
This would allow for permutations which are otherwise equivalent to be put through expensive operations. For example:

```
  A B C
A 0 1 2
B 1 0 3
C 2 3 0

...is equivalent too:

  A C B
A 0 2 1
C 2 0 3
B 1 3 0
```

The use of `.filter` here is in effect a reorder operation, not a filter. However, changing it so that the conditional is `set(self._ids) == set(ids)` would violate the return assumptions of order w.r.t. `ids` with implications to consumers. 

Would it make sense to add in a new kwarg to resolve whether order matters here?

Reply:
It is certainly possible to make it smarter.
However, checking for identity is very cheap and this patch tries to optimize for that only.
Additional tests would either be more expensive or potentially dangerous.

Note: tuple(['A','B','C'])!=tuple(['A','C','B']), so the current patch is semantically correct.
(even if I agree it could potentially be further improved)